Thread support for musl

dragosargint opened the pull request:
This PR contains essential changes for thread support in musl. This PR is currently a superset of https://github.com/unikraft/lib-musl/pull/5 , which will need to be merged first (So please don't bother with the first 4 commits). This PR is dependent on:
[1] Simon's uksched refactor: https://github.com/skuenzer/unikraft/tree/skuenzer/sched-refactor
[2] [unikraft/unikraft#454](https://github.com/unikraft/unikraft/pull/454)
[3] [unikraft/unikraft#453](https://github.com/unikraft/unikraft/pull/453)
[4][ unikraft/unikraft#442](https://github.com/unikraft/unikraft/pull/442)
[5][ unikraft/unikraft#441](https://github.com/unikraft/unikraft/pull/441)
[6] https://github.com/unikraft/unikraft/pull/488

[4] Cannot be automatically be merged into [1] right now, I will provide a patch on discord

dragosargint committed:
Add library dependencies for thread support

Select (by default) the core libraries used by musl:
`LIBUKMMAP` - provides mmap() syscall
`LIBPOSIX_PROCESS_PIDS` - provides pid,tid implementation
`LIBPOSIX_PROCESS_CLONE` - provides the clone() syscall
`LIBPOSIX_FUTEX` - provides futex() syscall
`LIBUKSCHED_TCB_INIT` - allows setting the size for the LIBC TCB

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Initialize TLS and reserve TCB space

This commit adds a call in Makefile.uk to reserve the space
needed for the LIBC TCB (e.g. `pthread` structure)
and adds 3 glue code files:
`__uk_init_tls.c` - initialize an Unikraft TLS
`__uk_unmapself.c` - allows unmapping for detached threads
`__set_thread_area.c` - bypass for arch_prctl() syscall for
setting the tls pointer

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Remove `__unmapself` and `set_thread_area`

Comment the sources corresponding to the functions
__unmapself() (for detached threads) and
__set_thread_area() (for setting the fs register
to the tls pointer) because they will be reintroduced
as glue code.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Provide `__set_thread_area()` as glue code

`__set_thread_area()` is a function that is usually
architecture dependent, it sets the tls pointer in the
fs register. An `arch_prctl` system call is used. We
are replacing this call with a custom version because
`arch_prctl()` is not implemented in Unikraft at the
moment.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Add custom `unmapself` function as glue code

`unmapself()` provides a detached thread the ability
to free the memory region allocated for the stack and
TLS. In the original version, this function does
`munmap()', which will deallocate the thread's stack,
after which it executes exit() (NB: SYS_exit). This
works because there is another stack in the Kernel.
Since this does not exist in Unikraft, we implement
our version of unmapself.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Add custom TLS related functions as glue code

The functions related to the TLS layout (like copying content)
are implemented in the unikraft core libraries.
To use them, we need to add glue source code.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Add patch for the `clone()` wrapper for x86_64

The clone wrapper is an assembly source file which will
invoke a `syscall` instruction. Since we want just a
function call, we need to change it accordingly.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
pthread_create: TLS alignment and cleanup

This patch replaces 2 function calls from musl with
2 custom ones and introduces an additional size
when allocating the map for a thread, `libc.tls_align`.
It is needed because Unikraft demands that the tls be aligned.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

dragosargint committed:
Remove obsolete patches

This commit reorders the patch numbers used for
musl and deletes those that are not needed as a
result of the changes made in this PR.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>

craciunoiuc commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,195 @@
+/*
+ * This glue code source is meant to replace `__libc_start_main()`
+ * and tls related stuff.
+ * If the pthread API is usead, musl will allocate the TLS and will
```
```suggestion
 * If the pthread API is used, musl will allocate the TLS and will
```

craciunoiuc commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,195 @@
+ * and tls related stuff.
+ * If the pthread API is usead, musl will allocate the TLS and will
+ * provide a pointer to it in the `clone()` syscall. Musl will allocate
+ * in fact a conigous region which will have the following layout (Please
```
```suggestion
 * in fact a contiguous region which will have the following layout (Please
```

craciunoiuc commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,195 @@
+	libc.can_do_threads = 1;
+	/*
+	 * The original musl code will invoke here a `SYS_set_tid_address`
+	 * syscall, to set the tid user space addres in the Kernel.
```
```suggestion
	 * syscall, to set the tid user space address in the Kernel.
```

craciunoiuc commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,195 @@
+{
+	struct pthread *td = (struct pthread *) tcb;
+
+	uk_pr_debug("uk_thread_tcb_fini uk_thraed %p, tcb %p\n", thread, tcb);
```
```suggestion
	uk_pr_debug("uk_thread_tcb_fini uk_thread %p, tcb %p\n", thread, tcb);
```

craciunoiuc reviewed (commented):
I did a first pass through all the files/commits and only found a few typos.

One mention: some of the commits have a pretty long commit title, try reducing them if possible.

Other than this, good work! I'll come back after testing it.

craciunoiuc commented:
All good from my side now, waiting on the other reviewers to add the tag.

flpostolache commented:
The PR looks good. I just have one question. Should we write somewhere that in order to make the clone wrapper function work, `CLONE_VM`, `CLONE_FILES` and `CLONE_FS` should be added as flags?

dragosargint commented:
> The PR looks good. I just have one question. Should we write somewhere that in order to make the clone wrapper function work, `CLONE_VM`, `CLONE_FILES` and `CLONE_FS` should be added as flags?

The flags are not added, they are default in musl for creating threads as we discussed. Indeed, in the release documentation, it should be mentioned that the `clone()' syscall only works for creating threads. Therefore, these flags must be used.

flpostolache commented:
Looks good.
Reviewed-by: Florin Postolache <florin.postolache80@gmail.com>

kubanrob commented:
It seems like `Makefile.uk.musl.process:#LIBMUSL_PROCESS_SRCS-y += $(LIBMUSL)/src/process/waitpid.c` needs to be commented in to build with the current Unikraft branch.

razvanvirtan commented:
> 

Yes, it happened to me too (I think you mean uncommented?)

dragosargint commented:
> It seems like `Makefile.uk.musl.process:#LIBMUSL_PROCESS_SRCS-y += $(LIBMUSL)/src/process/waitpid.c` needs to be commented in to build with the current Unikraft branch.

Hi, Robert! Thank you for your comment! You are right, that source must be uncommented from the Makefile. I modified this PR https://github.com/unikraft/lib-musl/pull/5 and rebased this one as well.

kubanrob commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,195 @@
+	 * The original musl code will invoke here a `SYS_set_tid_address`
+	 * syscall, to set the tid user space address in the Kernel.
+	 * FIXME: Currently this does not return the tid assigned for the caller,
+	 * it returns an error code (-95) because probably there is no tid assigned
+	 * at this stage. It is not a really big problem right now.
+	 */
+	td->tid = uk_syscall_r_set_tid_address(&td->tid);
```
This effectively sets the tid of the main thread to -95. This tid is then used, in the implementation of recursive mutexes, causing `pthread_mutex_unlock` to fail.

```suggestion
	 * it returns an error code (-95, -ENOTSUP) because posix_process_init has not been
	 * called at this stage, but will be called via uk_late_initcall.
	 * It is not a really big problem right now. Since this is the main thread,
	 * nobody should ever wait for it, and we can just assume thread id 0.
	 */
	td->tid = uk_syscall_r_set_tid_address(&td->tid);
	if (td->tid < 0) {
		td->tid = 0;
	}
```

dragosargint replied:
Hello, Robert. Thank you for your comment. You are right, the tid can be set to 0 here.

dragosargint committed:
Set tid to 0 for the main thread

This commit ads in the Config.uk the following default selection:
`LIBPOSIX_PROCESS_INIT_PIDS`, which allows pid,tid assignation
However, there is a bug in the assignment of tids that
makes the tid of the main thread not available when the
LIBC TCB (i.e. pthread structure) is initialized. The
currently adopted solution is to hardcode the tid to 0.

Signed-off-by: Dragos Iulian Argint <dragosargint21@gmail.com>
Signed-off-by: Robert Kuban <robert.kuban@opensynergy.com>

kubanrob commented:
@dragosargint Can you please change my information to `Robert Kuban <robert.kuban@opensynergy.com>` in commit 2a3c2a0? Yours are generated by the web interface as well.

razvand commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,200 @@
+	 * In the original code of musl this will use an `arch_prtcl`
+	 * syscall to fill the `$fs` register.
+	 */
+	ukplat_tlsp_set(TP_ADJ(p));
```
```suggestion
	ukplat_tlsp_set((unsigned long) TP_ADJ(p));
```

To suppress a warning.

dragosargint replied:
Thank you for your comment!

dragosargint commented:
> @dragosargint Can you please change my information to `Robert Kuban <robert.kuban@opensynergy.com>` in commit [2a3c2a0](https://github.com/unikraft/lib-musl/commit/2a3c2a0ae6dd4f1d1fa48af30f64eef110be0559)? Yours are generated by the web interface as well.

kubanrob commented:
Just 2 minor things:
- Space missing in commit message of eef2f0e03bf7be9ebeafa14376cf08e4ec97ba01: "is used.We are"
- Patch 15 is duplicated in 1e83a3ece0428a172fc71709f63229d66ccf535b just to be deleted in 46f65b5932832424c239b99089b9a6fc5a7fc659

craciunoiuc reviewed (commented):
Two small things from me 🙈

`lib-musl: remove no longer required patch` -> `lib-musl: Remove no longer required patch`
`Set tid to 0 for the main thread` -> `lib-musl: Set tid to 0 for the main thread`

razvand commented on Makefile.uk:
```diff
@@ -32,6 +32,14 @@
 #
 
+################################################################################
+# This will reserve space within `uk_tls` for the pthread structure in musl
```
```suggestion
# This will reserve space within `uk_tls` for the pthread structure in Musl.
```

razvand commented on Makefile.uk:
```diff
@@ -32,6 +32,14 @@
+################################################################################
+# This will reserve space within `uk_tls` for the pthread structure in musl
+# Please refer to ./arch/Makefile.rules fore more details.
+# For this musl version the size of "struct pthread" is 280
```
```suggestion
# For this version of Musl the size of "struct pthread" is 280.
```

razvand commented on Makefile.uk.musl.thread:
```diff
@@ -156,27 +156,27 @@ LIBMUSL_THREAD_SRCS-y += $(LIBMUSL)/src/thread/tss_set.c
 ifeq (x86_32,$(CONFIG_UK_ARCH))
-LIBMUSL_THREAD_SRCS-y += $(LIBMUSL)/src/thread/i386/__set_thread_area.s|i386
-LIBMUSL_THREAD_SRCS-y += $(LIBMUSL)/src/thread/i386/__unmapself.s|i386
+#LIBMUSL_THREAD_SRCS-y += $(LIBMUSL)/src/thread/i386/__set_thread_area.s|i386
```
Shouldn't the commented out lines simply be removed?

dragosargint replied:
There are several such lines commented out in the Makefiles of Musl. From my point of view, it is better to leave them commented out to see what is used from the real sources and what is not, or defined as glue code. That's why I just commented and not removed them.

razvand commented on __set_thread_area.c:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Authors: Dragos Iulian Argint <dragosargint21@gmail.com>
+ *
+ * Copyright (c) 2022, University Politehnica of Bucharest. All rights reserved.
```
```suggestion
 * Copyright (c) 2022, University POLITEHNICA of Bucharest. All rights reserved.
```

razvand commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,200 @@
+/*
+ * Authors: Dragos Iulian Argint <dragosargint21@gmail.com>
+ *
+ * Copyright (c) 2022, University Politehnica of Bucharest. All rights reserved.
```
```suggestion
 * Copyright (c) 2022, University POLITEHNICA of Bucharest. All rights reserved.
```

razvand commented on __uk_unmapself.c:
```diff
@@ -0,0 +1,85 @@
+/*
+ * Authors: Dragos Iulian Argint <dragosargint21@gmail.com>
+ *
+ * Copyright (c) 2022, University Politehnica of Bucharest. All rights reserved.
```
```suggestion
 * Copyright (c) 2022, University POLITEHNICA of Bucharest. All rights reserved.
```

razvand reviewed (changes requested):
Hi, @dragosargint . See some cosmetic comments from my side.

skuenzer commented on Config.uk:
```diff
@@ -4,9 +4,15 @@ menuconfig LIBMUSL
   default n
   select HAVE_LIBC
   select LIBKUALLOC
+  select LIBUKMMAP
```
Is mmap still used for the threading? If not, can we remove this dependency?

dragosargint replied:
We still need it :))

skuenzer commented on __uk_init_tls.c:
```diff
@@ -116,10 +116,15 @@ static int __uk_init_tp(void *p)
 	 */
 	td->tid = uk_syscall_r_set_tid_address(&td->tid);
+	if (td->tid < 0) {
+		td->tid = 0;
```
You may want to add to this commit the Config.uk option is selected that initializes the main thread id to 0.

skuenzer reviewed (changes requested):
Hey,
Dragos, please check my comments inline. I assume ukmmap is in use because of the musl code, right? If so, ignore my comment regarding the dependency.

Besides the left comments, I would change the subject of the following patches:
`Add patch for pthread_create.c` -> `pthread-create: TLS alignment and cleanup`
`Delete and rename some patches` -> `Remove obsolete patches`

Other from this, I think we are good to go.

Thanks,

Simon

dragosargint commented:
> Hi, @dragosargint . See some cosmetic comments from my side.

Thank you for the comments, @razvand. I applied your suggestions.

kubanrob commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,200 @@
+	libc.page_size = __PAGE_SIZE;
+}
+
+/*
```
This comment should probably go above the function definition.

dragosargint replied:
Yes, it should. I'll modify it

kubanrob commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,200 @@
+ * This callback will only be called for threads that are NOT
+ * created with the pthread API
+ */
+static volatile size_t __uk_tsd_size = sizeof(void *) * PTHREAD_KEYS_MAX;
```
Is there a reason that this is a static volatile variable? I don't see it referenced elsewhere and it looks like it should be an constant.

dragosargint replied:
I was actually copying it from musl code. I think we can remove the volatile keyword
```
volatile size_t __pthread_tsd_size = sizeof(void *) * PTHREAD_KEYS_MAX;
```

kubanrob commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,200 @@
+	return td;
+}
+
+static int __uk_init_tp(void *p)
```
I think a comment that says that this function should only be called for the main thread would help understanding this code. (if that is true.)

dragosargint replied:
It is true. It is the initialisation of the `pthread_structure` for the main thread. The other threads (created with the pthread API) will have this structure populated by the `pthread_create()` function.

kubanrob commented on __uk_init_tls.c:
```diff
@@ -0,0 +1,200 @@
+	 */
+	if (!libc.can_do_threads) {
+		__uk_init_tls(tcb);
+		return 0;
```
`__uk_init_tls` is only setting a subset of the tcb members set later in the function (eg. `td->tsd`). Is that on purpose?

dragosargint replied:
The `uk_thread_uktcb_init()` callback should be called for all threads that are not created via `clone()`. This means TLS allocated by unikraft, not by libc and passed through `clone()` to unikraft. The main thread is not created via `clone()` so `uk_thread_uktcb_init()` is called . `__uk_init_tls()` is called only once in the entire application and is only for the main thread, because after it executes it sets the libc.can_do_threads to 1.  What is below this if is for threads that are not created with `clone()' but that we want to run with musl.

dragosargint replied:
Here is the __init_tp for the main thread in musl
```
int __init_tp(void *p)
{
        pthread_t td = p;
        td->self = td;
        int r = __set_thread_area(TP_ADJ(p));
        if (r < 0) return -1;
        if (!r) libc.can_do_threads = 1;
        td->tid = __syscall(SYS_set_tid_address, &td->tid);
        td->locale = &libc.global_locale;
        td->robust_list.head = &td->robust_list.head;
        return 0;
}
```

dragosargint replied:
And you might wonder why there's no tsd for the main thread. Well, there is but is set on demand :))
```
int __pthread_key_create(pthread_key_t *k, void (*dtor)(void *))
{
...
        if (!self->tsd) self->tsd = __pthread_tsd_main;
}
```

dragosargint replied:
I now see that `uk_thread_uktcb_init()` is not called for the main thread. So `__uk_init_tls()` should be moved from there. But the other points are valid.

kubanrob reviewed (commented):
Looking at the code again, I found some things that are not clear to me.

I think a few comments what function is indented to be called from where would help making the glue code more understandable.

craciunoiuc commented on Makefile.uk:
```diff
@@ -32,6 +32,14 @@
 
+################################################################################
+# This will reserve space within `uk_tls` for the pthread structure in Musl.
+# Please refer to ./arch/Makefile.rules fore more details.
```
```suggestion
# Please refer to ./arch/Makefile.rules for more details.
```

craciunoiuc commented on __set_thread_area.c:
```diff
@@ -0,0 +1,40 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <uk/sched.h>
+/*
```
```suggestion
 */
 
#include <uk/sched.h>

/*
```

craciunoiuc commented on __uk_unmapself.c:
```diff
@@ -0,0 +1,85 @@
+	__munmap(m->map_base, m->map_size);
+}
+
+static __uk_tls struct __map_area __per_thread_map_area = {};
+void __uk_unmapself(void *base, size_t size)
```
```suggestion
static __uk_tls struct __map_area __per_thread_map_area = {};

void __uk_unmapself(void *base, size_t size)
```

craciunoiuc reviewed (commented):
Some small styling comments from my side

skuenzer reviewed (approved):
Thanks @dragosargint, the updates look good!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

razvand reviewed (approved):
Tested it, it works.

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>

kubanrob reviewed (approved):
Looks good, thank you @dragosargint 

Reviewed-by: Robert Kuban <robert.kuban@opensynergy.com>
